### ISSUE TITLE
`hh flatten` doesn't support circular dependencies

### ISSUE THREAD

**Issue by fvictorio** [nomiclabs/hardhat#1486]

**Comment by Endlessline** [comment]
How can one recreate this bug?

**Comment by fvictorio** [comment]
```solidity
// contracts/Foo.sol
import "./Bar.sol";

contract Foo {}
```

```solidity
// contracts/Bar.sol
import "./Foo.sol";

contract Bar {}
```

`hh compile` works, `hh flatten` doesn't.

The solution is probably to get some topological order of the dependency graph and to use that to generate the flattened file, ignoring previously visited files when resolving `import`s.

**Comment by BitMaster001** [comment]
+1

**Comment by nekooei** [comment]
+1

**Comment by slvrfn** [comment]
+1

**Comment by vinkent420** [comment]
+1

**Comment by nxqbao** [comment]
+1

**Comment by alcuadrado** [comment]
I updated this issue explaining why this is more complicated than it seems, and why it may still not work for some usecases. 

This is something low priority to us, but we may consider doing it once Slang is more mature.

**Comment by liu-zhipeng** [comment]
+1

**Comment by gnarvaja** [comment]
+1

**Comment by danilaplee** [comment]
+1

**Comment by aaitor** [comment]
+1

**Comment by jadrvel** [comment]
+1

**Comment by IsaacCisnerosFT** [comment]
+1

**Comment by xgamermonster9** [comment]
Yes, there’s some kind of givеаway going on. I got some Ethereum there yesterday for different wallets. It’s unlikely they’ll take it back from me, but it’s a pity I didn’t have more wallets and couldn’t get more. But maybe it’s for the best.

<img width="369" alt="Image" src="https://github.com/user-attachments/assets/5b983788-1d9d-45df-9185-dfb9584be220" />

<img width="373" alt="Image" src="https://github.com/user-attachments/assets/567fde7a-2777-4d4b-a13e-fd4c7fae960e" />

**Comment by donotyue** [comment]
+1

**Comment by adamcanray** [comment]
+1
